engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,848 @@
+	final
+)
+
+// TableAgent defines an interface for checkpoint.
+type TableAgent interface {
+	FetchAllDoTables(ctx context.Context, cfg *config.JobCfg) (map[metadata.TargetTable][]metadata.SourceTable, error)
+	FetchTableStmt(ctx context.Context, jobID string, cfg *config.JobCfg, sourceTable metadata.SourceTable) (string, error)
+}
+
+// DDLCoordinator is a coordinator for ddl.
+type DDLCoordinator struct {
+	mu          sync.Mutex
+	tables      map[metadata.TargetTable]map[metadata.SourceTable]struct{}
+	shardGroups map[metadata.TargetTable]*shardGroup
+	pendings    sync.WaitGroup
+	logger      *zap.Logger
+
+	kvClient   metaModel.KVClient
+	tableAgent TableAgent
+	jobID      string
+	jobStore   *metadata.JobStore
+}
+
+// NewDDLCoordinator creates a new DDLCoordinator.
+func NewDDLCoordinator(jobID string, kvClient metaModel.KVClient, tableAgent TableAgent, jobStore *metadata.JobStore, pLogger *zap.Logger) *DDLCoordinator {
+	return &DDLCoordinator{
+		tableAgent:  tableAgent,
+		tables:      make(map[metadata.TargetTable]map[metadata.SourceTable]struct{}),
+		jobID:       jobID,
+		kvClient:    kvClient,
+		shardGroups: make(map[metadata.TargetTable]*shardGroup),
+		jobStore:    jobStore,
+		logger:      pLogger.With(zap.String("component", "ddl_coordinator")),
+	}
+}
+
+// ClearMetadata clears metadata.
+func (c *DDLCoordinator) ClearMetadata(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// wait all pending ddls coordinated.
+	c.pendings.Wait()
+
+	return metadata.DelAllDroppedColumns(ctx, c.kvClient)
+}
+
+// Reset resets the ddl coordinator.
+func (c *DDLCoordinator) Reset(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// wait all pending ddls coordinated.
+	c.pendings.Wait()
+
+	jobCfg, err := c.jobStore.GetJobCfg(ctx)
+	if err != nil {
+		return err
+	}
+	if jobCfg.ShardMode == "" {
+		c.logger.Info("non-shard-mode, skip reset")
+		return nil
+	}
+	c.logger.Info("reset ddl coordinator")
+
+	// fetch all tables which need to be coordinated.
+	tables, err := c.tableAgent.FetchAllDoTables(ctx, jobCfg)
+	if err != nil {
+		return err
+	}
+
+	for targetTable, sourceTables := range tables {
+		c.tables[targetTable] = make(map[metadata.SourceTable]struct{}, 0)
+		for _, sourceTable := range sourceTables {
+			c.tables[targetTable][sourceTable] = struct{}{}
+		}
+	}
+	return nil
+}
+
+// Coordinate coordinates ddls.
+func (c *DDLCoordinator) Coordinate(ctx context.Context, item *metadata.DDLItem) ([]string, optimism.ConflictStage, error) {
+	c.mu.Lock()
+	c.pendings.Add(1)
+	defer c.pendings.Done()
+	c.mu.Unlock()
+
+	jobCfg, err := c.jobStore.GetJobCfg(ctx)
+	if err != nil {
+		return nil, optimism.ConflictError, err
+	} else if jobCfg.ShardMode == "" {
+		return nil, optimism.ConflictError, errors.New("coordinate error with non-shard-mode")
+	}
+
+	for {
+		// create shard group if not exists.
+		g, err := c.loadOrCreateShardGroup(ctx, item.TargetTable, jobCfg)
```

Comment:
concurrent support for `Coordinate` brings many troubles, we need extra logic for protecting c.pending, g.deleted,  for-loop to optimistically imitate no-locking, maybe we should use simple implementation next time.

Reply:
Unless we use distributed locks, there doesn't seem to be a particularly good way to do it

Reply:
Maybe we can change to synchronous in the future.

Reply:
[2562bd7](https://github.com/pingcap/tiflow/pull/7523/commits/2562bd758df1bd386391d8fde07b75f5f747c7a0) synchronous